byterun/win32.c

```diff
@@ -914,3 +918,62 @@ void caml_restore_win32_terminal(void)
   if (startup_codepage != 0)
     SetConsoleOutputCP(startup_codepage);
 }
+
+/* Detect if a named pipe corresponds to MSYS/Cygwin tty: see
+
+   https://github.com/mirror/newlib-cygwin/blob/master/winsup/cygwin/dtable.cc#L932
+*/
+typedef
+BOOL (WINAPI *tGetFileInformationByHandleEx)(
+  HANDLE                    hFile,
+  FILE_INFO_BY_HANDLE_CLASS FileInformationClass,
+  LPVOID                    lpFileInformation,
+  DWORD                     dwBufferSize
+);
+
+static int caml_win32_ismsystty(HANDLE hFile)
+{
+  char buffer[1024];
+  FILE_NAME_INFO * nameinfo = (FILE_NAME_INFO *) buffer;
+  HMODULE hModKernel32;
+  tGetFileInformationByHandleEx pGetFileInformationByHandleEx;
+
+  /* check if fd is a pipe */
+  if (GetFileType(hFile) != FILE_TYPE_PIPE)
+    return 0;
+
+  hModKernel32 = GetModuleHandle(L"KERNEL32.DLL");
+  pGetFileInformationByHandleEx =
+    (tGetFileInformationByHandleEx)GetProcAddress(hModKernel32, "GetFileInformationByHandleEx");
+
+  if (pGetFileInformationByHandleEx == NULL)
+    return 0;
+
+  /* get pipe name */
+  if (! pGetFileInformationByHandleEx(hFile, FileNameInfo, buffer, sizeof(buffer) - sizeof(WCHAR)))
+    return 0;
+
+  nameinfo->FileName[nameinfo->FileNameLength / sizeof(WCHAR)] = L'\0';
```

Comment:
I am passing the maximum buffer size `(sizeof(buffer) - sizeof(WCHAR))` to the `GetFileInformationByHandleEx` call.  Do I understand correctly that you want the bounds check in case the system call returns something erroneous?

Reply:
Ah, I'm sorry - I'd missed the subtlety of the `- sizeof(WCHAR)` ... please could that be commented on the call (something like "GetFileInformationByHandleEx doesn't null-terminate the string, so reduce the declared buffer length to allow for adding one."